core/src/main/java/org/jdbi/v3/core/collector/OptionalBuilder.java

```diff
@@ -13,22 +13,51 @@
  */
 package org.jdbi.v3.core.collector;
 
-import java.util.Optional;
-
-class OptionalBuilder<T> {
-    private Optional<T> optional = Optional.empty();
-
-    public void set(T value) {
-        if (optional.isPresent()) {
-            throw new IllegalStateException(
-                    String.format("Multiple values for Optional type: ['%s','%s',...]",
-                            optional.get(),
-                            value));
-        }
-        optional = Optional.ofNullable(value);
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+class OptionalBuilder<T, OPT_T> {
```

Comment:
just my 2 cents here: maybe `OptionalBuilder` is not the best class name since it kind of implies the java8 optional. basically it is a builder for all zero-one collections.
also OPT_T reads a little weird to me. maybe stick with R as in the java8 collector api (where R is the return value type)

Reply:
The class name thing was on my mind when I was working on this. There are several different terms to choose from just looking at the [Wikipedia page on the topic](https://en.wikipedia.org/wiki/Option_type). Ultimately I decided to leave it as `OptionalBuilder` since `Optional` is the term the JDK uses, and this builder is package private anyway.

As for the generic parameter name, I've actually names like this in several JDK classes, e.g. `PrimitiveIterator<T, T_CONS>`, `AbstractPipeline<E_IN, E_OUT, S>`, along with several other types in the `java.util.stream` package. That said, we can probably get away with renaming this parameter to just `O`.